coding/python_testing.rst

```diff
@@ -42,44 +41,149 @@ The important things to note in this example are:
 
+The tests/SConscript file
+-------------------------
+
+The behavior of `pytest`_ when invoked by :command:`scons` is controlled by the :file:`tests/SConscript` file.
+At minimum this file should contain the following to enable testing with automated test discovery:
+
+.. code-block:: python
+
+   from lsst.sconsUtils import scripts
+   scripts.BasicSConscript.tests(pyList=[])
+
+``pyList`` is used to specify which Python test files to run.
+Here the empty list is interpreted as "allow :command:`pytest` to automatically discover tests"
+:command:`pytest` will scan the directory tree itself to find tests and will run them all in together using the number of subprocesses matching the ``-j`` argument given to :command:`scons`.
+For this mode to work, all test files must be named :file:`test_*.py`.
+
+If ``pyList=None`` (the historical default) is used, the :command:`scons` ``tests`` target will be used to locate test files using a glob for ``*.py`` in the :file:`tests` directory.
+This list will then be passed explicitly to the :command:`pytest` command, bypassing its automatic test discovery.
+
+Automatic test discovery is preferred as this ensures that there is no difference between running the tests with :command:`scons` and running them with :command:`pytest` without arguments, and it enables the possibility of adjusting :command:`pytest` test discovery to add additional testing of all Python files in the package.
+
+Running tests standalone
+------------------------
+
+``pySingles`` is an optional argument to the :lmeth:`~lsst.sconsUtils.scripts.BasicSConscript.tests` method that can be used for the rare cases where a test must be run standalone and not with other test files in a shared process.
+
+.. code-block:: python
+
+   scripts.BasicSConscript.tests(pyList=[], pySingles=["testSingle.py"])
+
+The tests are still run using :command:`pytest` but executed one at a time without using any multi-process execution.
+Use of this should be extremely rare.
+In the :lmod:`~lsst.base` package one test file is used to confirm that the LSST import code is working; this can only be tested if we know that it hasn't previously been imported as part of another test.
+The other reason, so far, to run a test standalone is for test classes that dynamically generate large amounts of test data during the set up phase.
+Until it is possible to pin test classes to a particular process with ``pytest-xdist``, tests such as these interact badly when test methods within the class are allocated to different subprocesses since each subprocess will generate the test files.
+This can use significantly more disk and CPU when the test runs, and can even cause Jenkins to fail.
+It is important to ensure that any files listed in ``pySingles`` should be named such that they will not be discovered by `pytest`_.
+The convention is to name these files :file:`test*.py` without the underscore.
```

Comment:
For my own edification, does `pySingles` interact properly with `pyList=None`?

Reply:
Of course :-)  `daf_base` is already doing that.